share/src/test/java/org/apache/kafka/server/share/persister/DefaultStatePersisterTest.java

```diff
@@ -0,0 +1,558 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.server.share.persister;
+
+import org.apache.kafka.clients.KafkaClient;
+import org.apache.kafka.clients.MockClient;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.Uuid;
+import org.apache.kafka.common.message.FindCoordinatorResponseData;
+import org.apache.kafka.common.message.ReadShareGroupStateRequestData;
+import org.apache.kafka.common.message.ReadShareGroupStateResponseData;
+import org.apache.kafka.common.message.WriteShareGroupStateRequestData;
+import org.apache.kafka.common.protocol.Errors;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.FindCoordinatorResponse;
+import org.apache.kafka.common.requests.ReadShareGroupStateRequest;
+import org.apache.kafka.common.requests.ReadShareGroupStateResponse;
+import org.apache.kafka.common.requests.WriteShareGroupStateRequest;
+import org.apache.kafka.common.requests.WriteShareGroupStateResponse;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.server.share.SharePartitionKey;
+import org.apache.kafka.server.util.MockTime;
+
+import org.junit.jupiter.api.Test;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.mockito.Mockito.mock;
+
+class DefaultStatePersisterTest {
```

Comment:
Probably ought to test the error paths in the DefaultStatePersister too. For example, if you get an ExecutionException. Having done work in the admin client where the exception handling involving futures in the public API was not quite right, I'm keen to make sure we don't have this kind of mistake here too.

Reply:
@AndrewJSchofield 
For `DefaultStatePersister` there is no specific error path. This is because we are completing all futures normally in the `PersisterStateManager` - which actually sends the RPCs over the network - even if the RPC response has errors. 
In error case we complete the future with corresponding error code in `PersisterStateManager` but it is still completed normally.
This is necessary because we are using `allOf` to combine futures in `DefaultStatePersister`.
By definition, `allOf` returns a future which is completed when all underlying futures are completed. If any future completes exceptionally, the `allOf` just contains that exception and we lose partial results. To prevent this we complete futures in `PersisterStateManager` normally error or no error.